### PR TITLE
Master inventory report arm

### PR DESCRIPTION
**Pull request by amoyaux**

### CODE REVIEW

File: addons/stock_account/models/stock_quant.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class StockQuant(models.Model):
+    _inherit = 'stock.quant'
+
+    value = fields.Float('Value', compute='_compute_value')
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_quant.py]: probably missing a groups inventory manager

access right issue when inventory user> 
Sorry, you are not allowed to access documents of type 'Decimal Precision' (decimal.precision). This operation is allowed for the groups: - Administration/Settings - (Operation: read, User: 6)

File: addons/stock_account/models/stock_quant.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        for quant in self:
+            if quant.product_id.cost_method == 'fifo':
+                quantity = quant.product_id.quantity_svl
+                if not quantity:
+                    quant.value = 0.0
+                    continue
+                average_cost = quant.product_id.value_svl / quantity
+                quant.value = quant.quantity * average_cost
+            else:
+                quant.value = quant.quantity * quant.product_id.standard_price
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_quant.py]: maybe we could try reading value_svl / quantity_svl out of the loop
``` python
        fifo_products = self.env['product.product']
        standard_avco_products = self.env['product.product']
        for quant in self:
            if quant.product_id.cost_method == 'fifo':
                fifo_products |= quant.product_id
            else:
                standard_avco_products |= quant.product_id
        quantity_svl_per_product = {product: product.quantity_svl for product in fifo_products}
        value_svl_per_product = {product: product.value_svl for product in fifo_products}
        standard_price_per_product = {product: product.standard_price for product in standard_avco_products}

        for quant in self:
            if quant.product_id in fifo_products:
                quantity_svl = quantity_svl_per_product[quant.product_id]
                if float_is_zero(quantity_svl, precision_rounding=quant.product_id.rounding):
                    quant.value = 0
                    continue
                value_svl = value_svl_per_product[quant.product_id]
                quant.value = (value_svl / quantity_svl) * quant.quantity
            else:
                quant.value = quant.quantity * standard_price_per_product[quant.product_id]
```

it doesn't make a difference with the 18quants of the demo data though :(

File: addons/stock/models/stock_quant.py [hunk 3]
```diff
@@ -163,6 +163,30 @@ def action_view_stock_moves(self):
...
+        if not self.user_has_groups('stock.group_stock_multi_locations'):
+            company_user = self.env.user.company_id
+            warehouse = self.env['stock.warehouse'].search([('company_id', '=', company_user.id)], limit=1)
+            if warehouse:
+                self = self.with_context(default_location_id=warehouse.lot_stock_id.id)
+
+        # If user have rights to write on quant, we set quants in inventory mode.
+        if self.user_has_groups('stock.group_stock_manager'):
+            self = self.with_context(inventory_mode=True)
+        return self.env['stock.quant']._get_quants_action(extend=True)
```
Comment (sle-odoo) [inline addons/stock/models/stock_quant.py]: return self._get_quants_action(extend=True)

File: addons/stock/wizard/stock_quantity_history.py [hunk 4]
```diff
@@ -1,57 +1,29 @@
...
+        # We pass `to_date` in the context so that `qty_available` will be computed across
+        # moves until date.
+        action = {
+            'type': 'ir.actions.act_window',
+            'views': [(tree_view_id, 'tree'), (form_view_id, 'form')],
+            'view_mode': 'tree,form',
+            'name': _('Products'),
+            'res_model': 'product.product',
+            'domain': "[('type', '=', 'product')]",
+            'context': dict(self.env.context, to_date=self.inventory_datetime.date()),
```
Comment (sle-odoo) [inline addons/stock/wizard/stock_quantity_history.py]: cast to date?

File: addons/stock/wizard/stock_quantity_history.xml [hunk 5]
```diff
@@ -1,31 +1,18 @@
...
                 <group>
-                    <group>
-                        <field name="compute_at_date" widget="radio"/>
-                        <field name="date" attrs="{'invisible': [('compute_at_date', '=', '0')]}"/>
-                    </group>
+                    <field name="inventory_datetime"/>
                 </group>
                 <footer>
-                    <button name="open_table" string="Retrieve the Inventory Quantities" type="object" class="btn-primary"/>
+                    <button name="open_at_date" string="Confirm" type="object" class="btn-primary"/>
```
Comment (sle-odoo) [inline addons/stock/wizard/stock_quantity_history.xml]: i don't see where this view is referenced now that you removed the stock account action

File: addons/stock_account/views/stock_valuation_layer_views.xml [hunk 6]
```diff
@@ -41,20 +41,21 @@
...
-            <tree default_order="create_date desc, id desc" create="0" import="0">
+            <tree default_order="create_date desc, id desc" create="0"
+                  import="0" js_class="inventory_report_list">
                 <field name="create_date" string="Date" />
                 <field name="product_id" />
                 <field name="quantity" />
                 <field name="uom_id" groups="uom.group_uom" />
                 <field name="currency_id" invisible="1" />
-                <field name="value" />
+                <field name="value"/>
```
Comment (sle-odoo) [inline addons/stock_account/views/stock_valuation_layer_views.xml]: useless diff

File: addons/stock_account/views/stock_account_views.xml [hunk 7]
```diff
@@ -92,13 +91,17 @@
...
-            <field name="target">new</field>
-            <field name="context">{'default_compute_at_date': '0', 'valuation': True}</field>
+            <field name="model_id" ref="stock.model_stock_quantity_history"/>
+            <field name="state">code</field>
+            <field name="code">
+                action = model.create({
+                  'date': datetime.date.today()
+                }).with_context(
+                    active_model='stock.valuation.layer'
+                ).open_at_date()
```
Comment (sle-odoo) [inline addons/stock_account/views/stock_account_views.xml]: not needed here

File: addons/stock_account/models/stock_quant.py (hunk 2 above)
Comment (amoyaux) [reply addons/stock_account/models/stock_quant.py]: 1700ms for 7000 quants(1500 as FIFO) with this code 1400ms for same date without

File: addons/stock/wizard/stock_quantity_history.xml (hunk 5 above)
Comment (amoyaux) [reply addons/stock/wizard/stock_quantity_history.xml]: removed the declaration in stock_account

File: addons/stock_account/models/stock_quant.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class StockQuant(models.Model):
+    _inherit = 'stock.quant'
+
+    value = fields.Float('Value', compute='_compute_value', groups='stock.group_stock_manager')
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_quant.py]: could be a monetary field

File: addons/stock_account/views/stock_quant_views.xml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    </record>
+
+    <record model="ir.ui.view" id="view_stock_quant_tree_editable_inherit">
+        <field name="name">stock.quant.tree.editable.inherit</field>
+        <field name="model">stock.quant</field>
+        <field name="inherit_id" ref="stock.view_stock_quant_tree_editable"></field>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='product_uom_id']" position="after">
+                <field name="currency_id" invisible="1"/>
+                <field name="value" widget='monetary' options="{'currency_field': 'currency_id'}"/>
```
Comment (sle-odoo) [inline addons/stock_account/views/stock_quant_views.xml]: ```<field name="value" /> once a monetary field```

File: addons/stock/models/stock_quant.py [hunk 10]
```diff
@@ -163,6 +163,30 @@ def action_view_stock_moves(self):
         ]
         return action
 
+    def action_view_quants(self):
+        self = self.with_context(search_default_internal_loc=1)
+        if self.user_has_groups('stock.group_production_lot,stock.group_stock_multi_locations'):
+            # fixme: erase the following condition when it'll be possible to create a new record
+            # from a empty grouped editable list without go through the form view.
+            if self.env['stock.quant'].search_count([
```
Comment (sle-odoo) [inline addons/stock/models/stock_quant.py]: if self.search_count

File: addons/stock/models/stock_quant.py [hunk 11]
```diff
@@ -163,6 +163,30 @@ def action_view_stock_moves(self):
...
+                ('company_id', '=', self.env.user.company_id.id),
+                ('location_id.usage', 'in', ['internal', 'transit'])
+            ]) > 0:
```
Comment (sle-odoo) [inline addons/stock/models/stock_quant.py]: no need for > 0

File: addons/stock/models/stock_quant.py [hunk 12]
```diff
@@ -163,6 +163,30 @@ def action_view_stock_moves(self):
         ]
         return action
 
+    def action_view_quants(self):
+        self = self.with_context(search_default_internal_loc=1)
```
Comment (sle-odoo) [inline addons/stock/models/stock_quant.py]: do we need an ensure one ?

File: addons/stock_account/views/stock_valuation_layer_views.xml (hunk 6 above)
Comment (sle-odoo) [reply addons/stock_account/views/stock_valuation_layer_views.xml]: et mec

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r-

File: addons/stock_account/models/stock_quant.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        for quant in self:
+            if quant.product_id.cost_method == 'fifo':
+                quantity = quant.product_id.quantity_svl
+                if float_is_zero(quantity, precision_rounding=quant.product_id.uom_id.rounding):
+                    quant.value = 0.0
+                    continue
+                average_cost = quant.product_id.value_svl / quantity
+                quant.value = quant.quantity * average_cost
+            else:
+                quant.value = quant.quantity * quant.product_id.standard_price
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_quant.py]:
``` python
if not quant.location_id._should_be_valued() or\
        (quant.owner_id and quant.owner_id != self.env.company_id.partner_id):
    quant.value = 0
```

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**